sqlccl: add incremental progress to local csv import

madelynnblue opened the pull request:
This required adding a new Size method to ExportStorage, which works
for all storages except S3 (due to a client library limitation,
not S3).

Progress is divided into two 50% chunks. The first is for reading
CSVs into KVs. We use the new Size method to find the total size of
the CSV files, and track how many bytes we've read. If Size fails,
then we just report after reading each file. The second phase reports
the number of KVs written against the total number of KVs, which are
known by this time.

This means the before- and after-50% rates may differ, but that's ok
and may be as good as we can get, since it is unknown how long the
write SST phase will take, since it may depend on network speed.

Distributed will be in a future PR.

See #18630

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18677)
<!-- Reviewable:end -->

dt commented on pkg/ccl/sqlccl/csv.go:
```diff
@@ -338,7 +377,30 @@ func readCSV(
+		if err != nil {
+			return 0, err
+		}
+		sz, _ := es.Size(ctx, "")
```
do we want to log the error? or potentially also log a warning explaining that we'll fall back in the size == 0 block below?

dt commented on pkg/ccl/storageccl/export_storage.go:
```diff
@@ -268,11 +283,22 @@ func (h *httpStorage) Delete(_ context.Context, basename string) error {
 }
 
+func (h *httpStorage) Size(_ context.Context, basename string) (int64, error) {
+	resp, err := h.req("HEAD", basename, nil)
```
Some servers might not implement this -- at least it isn't part of our current required API.
We potentially want to add a test where our mock http server errors or otherwise returns an unusable response, just to make sure we keep handling it gracefully.

dt commented:
Some HTTP servers may or may not support the HEAD verb as this api expects, so we should potentially include a test where our mock server errors or returns garbage results (unless we add the HEAD verb to our required semantics) to ensure we handle it reasonably gracefully when we don't have usable size info.

---

Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/sqlccl/csv.go, line 449 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-KuwSA6Y7HPL3pF49IDS:-KuwSA6Y7HPL3pF49IDT:boh4ez0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e625f3a5fe66ec0acca1650aa39f5e201980686/pkg/ccl/sqlccl/csv.go#L449)):*
> ```Go
> 						}
> 					}
> 					const FiftyMiB = 50 << 20
> ```

capitalized?

---

*[pkg/ccl/storageccl/export_storage.go, line 379 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-KuwRD4eASUMp59KpWo5:-KuwRD4eASUMp59KpWo6:bd06r8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e625f3a5fe66ec0acca1650aa39f5e201980686/pkg/ccl/storageccl/export_storage.go#L379)):*
> ```Go
> 
> func (s *s3Storage) Size(_ context.Context, basename string) (int64, error) {
> 	// TODO(mjibson): https://github.com/rlmcpherson/s3gof3r/issues/124
> ```

yeah, we should potentially just switch to the official aws sdk in a follow up, if it now supports the features we need (streaming).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-:-KuwQdcXJqJuU1_CR_l_:b-ke8szh)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sqlccl: add incremental progress to local csv import

This required adding a new Size method to ExportStorage, which works
for all storages except S3 (due to a client library limitation,
not S3).

While here, removed some needless r.Body.Close() in
export_storage_test.go because Go closes server request bodies itself.

Progress is divided into two 50% chunks. The first is for reading
CSVs into KVs. We use the new Size method to find the total size of
the CSV files, and track how many bytes we've read. If Size fails,
then we just report after reading each file. The second phase reports
the number of KVs written against the total number of KVs, which are
known by this time.

This means the before- and after-50% rates may differ, but that's ok
and may be as good as we can get, since it is unknown how long the
write SST phase will take, since it may depend on network speed.

Distributed will be in a future PR.

See #18630

mjibson commented:
Added a test for 500 responses on HEAD requests.

---

Review status: 1 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/ccl/sqlccl/csv.go, line 392 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-Ku_ugb_G0In8WloAAUz-r1-392:-KuzKYtzG-LiHrgZbXR7:b242t0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e625f3a5fe66ec0acca1650aa39f5e201980686/pkg/ccl/sqlccl/csv.go#L392)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

do we want to log the error? or potentially also log a warning explaining that we'll fall back in the size == 0 block below?
</blockquote></details>

Added an info when we fallback.

---

*[pkg/ccl/sqlccl/csv.go, line 449 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-KuwSA6Y7HPL3pF49IDS:-KuzKaCRLd0JrmMLbkHG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e625f3a5fe66ec0acca1650aa39f5e201980686/pkg/ccl/sqlccl/csv.go#L449)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

capitalized?
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/export_storage.go, line 287 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-Ku_ugb_G0In8WloAAV--r1-287:-KuzKb7IWVVtBhwaNL_T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9e625f3a5fe66ec0acca1650aa39f5e201980686/pkg/ccl/storageccl/export_storage.go#L287)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Some servers might not implement this -- at least it isn't part of our current required API.
We potentially want to add a test where our mock http server errors or otherwise returns an unusable response, just to make sure we keep handling it gracefully.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-:-KuzKVlqpyjGR8bMQ9Pt:bit7041)*
<!-- Sent from Reviewable.io -->

mjibson commented:
@dt PTAL at the new HTTP tests.

dt commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18677#-:-KuzdqPwFdUVi_ZXE6ZP:bnfp4nl)*
<!-- Sent from Reviewable.io -->
